### PR TITLE
fix: added check for peer's pairs on order invalidation

### LINKED ISSUE #1530: xud processing order updates for pairs I do not have
### How it is

I do **not** have the pair btc/xuc set up on my xud (I do have xuc/btc set up though), xud is receiving order invalidations for such peer and actually tries to remove the order; which fails of course.

```
07/05/2020 11:24:11.754 [P2P] verbose: received order invalidation from 02778bcd8f1f578bbb2f3ff345ecccdb5865e4a08a0cfd98145da08ca3470c3e74 (ReadyAbsurd): {"id":"409fb610-9055-11ea-b76d-13690b06b3b3","pairId":"BTC/XUC","quantity":10000000}
07/05/2020 11:24:11.755 [ORDERBOOK] error: failed to remove order (409fb610-9055-11ea-b76d-13690b06b3b3) of peer 02778bcd8f1f578bbb2f3ff345ecccdb5865e4a08a0cfd98145da08ca3470c3e74 (ReadyAbsurd)
```

### How it should be

- xud knows about a peers pairs, don't send updates for pairs the peer doesn't have
- if an update gets send anyways, xud on the receiving side ignores such an update, adding a log entry in trace level, but not default log level (debug)

### PR DESCRIPTION
**Pull request by rsercano**
attempts to fix #1530 

Not sure if there's any other order event to check, also it would be great if you can tell me if I have to check pair's active status before emitting the event.

### CODE REVIEW

File: lib/p2p/Pool.ts [hunk 1]
```diff
@@ -728,7 +728,11 @@ class Pool extends EventEmitter {
     const orderInvalidationPacket = new packets.OrderInvalidationPacket({ id, pairId, quantity });
     this.peers.forEach(async (peer) => {
       if (!nodeToExclude || peer.nodePubKey !== nodeToExclude) {
-        await peer.sendPacket(orderInvalidationPacket);
+        if (peer.advertisedPairs.includes(pairId)) {
+          await peer.sendPacket(orderInvalidationPacket);
+        } else {
+          this.logger.trace(`skipping to send order invalidation to ${peer.address} since pair is not found`);
```
Comment (sangaman) [inline lib/p2p/Pool.ts:734]: I don't think we need this logging statement even on the trace level, it's basically a non-event.

File: lib/p2p/Pool.ts [hunk 2]
```diff
@@ -728,7 +728,11 @@ class Pool extends EventEmitter {
     const orderInvalidationPacket = new packets.OrderInvalidationPacket({ id, pairId, quantity });
     this.peers.forEach(async (peer) => {
       if (!nodeToExclude || peer.nodePubKey !== nodeToExclude) {
-        await peer.sendPacket(orderInvalidationPacket);
+        if (peer.advertisedPairs.includes(pairId)) {
```
Comment (sangaman) [inline lib/p2p/Pool.ts:731]: Better to use `peer.isPairActive(pairId)` here as sometimes advertised pairs are not active (and therefore we wouldn't be sending orders for that pair to the peer in the first place).

File: lib/p2p/Pool.ts [hunk 3]
```diff
@@ -786,8 +790,12 @@ class Pool extends EventEmitter {
       }
       case PacketType.OrderInvalidation: {
         const orderInvalidation = (packet as packets.OrderInvalidationPacket).body!;
-        this.logger.trace(`received order invalidation from ${peer.label}: ${JSON.stringify(orderInvalidation)}`);
-        this.emit('packet.orderInvalidation', orderInvalidation, peer.nodePubKey as string);
+        if (!this.nodeState.pairs.includes(orderInvalidation.pairId)) {
```
Comment (sangaman) [inline lib/p2p/Pool.ts:793]: Again here it's better to use `peer.isPairActive(pairId)`. Also minor nit but I think the code is a bit easier to read if the if statement checks for the true/active case first, then else prints the inactive pair message.

File: lib/p2p/Pool.ts [hunk 4]
```diff
@@ -786,8 +790,12 @@ class Pool extends EventEmitter {
...
+          this.logger.trace(`received order invalidation for unrelated pair from ${peer.label}: ${JSON.stringify(orderInvalidation)}`);
```
Comment (sangaman) [inline lib/p2p/Pool.ts:794]: "inactive" rather than "unrelated" to be consistent with wording elsewhere.

**Review by sangaman — CHANGES_REQUESTED** [review]
Thanks for addressing this, it's a minor fix but makes sense and will make the logs cleaner.

File: lib/p2p/Pool.ts (hunk 3 above)
Comment (rsercano) [reply lib/p2p/Pool.ts:793]: But isn't in this case we'll check incoming peer's active pairs instead of our active pairs?

**Comment by raladev** [comment]
```
04/09/2020 13:07:09.941 [P2P] verbose: received node state update packet from 03bdf0e1b1338711cd4374802b6467793fe455f3c7adfd67935e1822c7d12cfc16 (DawnAerobic): {"pairs":["ETH/BTC","LTC/BTC","LTC/USDT","BTC/USDT","USDT/ETH"],"addresses":[{"host":"7w6fk4s6tvpsnesw55voeluvruv3kxtbgdk3gacydcl6wekndns5x7ad.onion","port":28885}],"connextIdentifier":"indra6Wt45v28puTyh5A5dpLJmW8kNfss8vQeS5zGCLNWSUY4tDHDfg","lndPubKeys":{"BTC":"02cbeda716bb4d455f4d45decac5f58b47f46bd2b8dd15951ed63c5413065a697d","LTC":"02cbeda716bb4d455f4d45decac5f58b47f46bd2b8dd15951ed63c5413065a697d"},"lndUris":{"BTC":["02cbeda716bb4d455f4d45decac5f58b47f46bd2b8dd15951ed63c5413065a697d@ue36scvguve6rthvrdwxjsn4fru266ggyvyamotmqrzpgt24p3ip3hqd.onion:29735"],"LTC":["02cbeda716bb4d455f4d45decac5f58b47f46bd2b8dd15951ed63c5413065a697d@wuqkkmuthtmzlvity5g2mwvcqgvccuxyta4asadrldgd6z4exlnya7ad.onion:30735"]},"tokenIdentifiers":{"BTC":"bitcoin-simnet","ETH":"0x0000000000000000000000000000000000000000","LTC":"litecoin-simnet","USDT":"0x61F6b6D86E29fC62224943583F97966F093cC94D"}}


04/09/2020 13:07:09.942 [P2P] trace: Sent Get Orders packet to 03bdf0e1b1338711cd4374802b6467793fe455f3c7adfd67935e1822c7d12cfc16 (DawnAerobic): "{\"body\":{\"pairIds\":[\"USDT/ETH\"]},\"header\":{\"id\":\"8a8d3960-eeaf-11ea-8df1-fd04cc7ae94a\"}}"



04/09/2020 13:07:10.471 [P2P] trace: Received Orders packet from 03bdf0e1b1338711cd4374802b6467793fe455f3c7adfd67935e1822c7d12cfc16 (DawnAerobic): "{\"body\":[],\"header\":{\"id\":\"8abfe220-eeaf-11ea-8d17-c395fd7a8440\",\"reqId\":\"8a8d3960-eeaf-11ea-8df1-fd04cc7ae94a\"}}"
04/09/2020 13:07:10.472 [P2P] verbose: received 0 orders from 03bdf0e1b1338711cd4374802b6467793fe455f3c7adfd67935e1822c7d12cfc16 (DawnAerobic)
```
- [ ] after adding new pair, the peer with new pair sends packet update to his peers. And after that these peers send GetOrders packet to this peer even if they have no this pair. It would be good to not ask if we have no this new pair.
 Should we create new issue or we can fix it it this PR? @rsercano @kilrau 

- [ ] Also it would be good to send GetOrders packet from peer that added pair (if remote peer has this pair). It will help to fix this issue - https://github.com/ExchangeUnion/xud/issues/1538

**Review by raladev — CHANGES_REQUESTED** [review]
above

File: lib/p2p/Pool.ts (hunk 3 above)
Comment (sangaman) [reply lib/p2p/Pool.ts]: Yes that is true, however I think that's what we want here. If the peer's pair isn't active, then we won't have any orders for it in the order book anyway and the invalidation is moot.

**Comment by sangaman** [comment]
>     * [ ]  Also it would be good to send GetOrders packet from peer that added pair (if remote peer has this pair). It will help to fix this issue - #1538

I would do this in a separate PR, although this is a good idea. Just don't want this one to get too big or complicated.

**Comment by rsercano** [comment]
```
after adding new pair, the peer with new pair sends packet update to his peers. And after that these peers send GetOrders packet to this peer even if they have no this pair. It would be good to not ask orders if we have no this new pair.
Should we create new issue or we can fix it it this PR? @rsercano @kilrau
```

Alight, I'll check this point within this PR.

**Comment by rsercano** [comment]
I added a new check as per @raladev's request, now we won't send NodeStateUpdate packet when adding/removing a pair if this pair is not active in the peer.

**Comment by raladev** [comment]
> I added a new check as per @raladev's request, now we won't send NodeStateUpdate packet when adding/removing a pair if this pair is not active in the peer.

Not sure that it is correct fix, we still want to get info about pairs that we dont have, but we dint need info about orders. So, more correct would be remove GetOrders call in this case

**Comment by rsercano** [comment]
> > I added a new check as per @raladev's request, now we won't send NodeStateUpdate packet when adding/removing a pair if this pair is not active in the peer.
> 
> Not sure that it is correct fix, we still want to get info about pairs that we dont have, but we dint need info about orders. So, more correct would be remove GetOrders call in this case

I see, let me check it this way then

**Comment by rsercano** [comment]
fixed

**Comment by raladev** [comment]
> fixed

but i still dont see the code for handling of GetOrders autocall on remote peer side.
It seems u just returned UpdatePacket without GetOrders handling.
Will test.

**Comment by rsercano** [comment]
I thought this would be sufficient https://github.com/ExchangeUnion/xud/pull/1864/files#diff-1934dba4817446d9f1708503332eb124R731

**Comment by raladev** [comment]
> I thought this would be sufficient https://github.com/ExchangeUnion/xud/pull/1864/files#diff-1934dba4817446d9f1708503332eb124R731

- [ ] still see GetOrders packet sending :(
Screenshot from 2020-09-07 17-09-48

**Review by raladev — CHANGES_REQUESTED** [review]
above

**Comment by rsercano** [comment]
Alright let me recheck

**Review by sangaman — APPROVED** [review]
This looks good to me. Let's handle the `GetOrders` packet in another PR, I think the change here can stand on its own and fixes the behavior with the order invalidation packets that are mentioned in the original issue.

**Comment by sangaman** [comment]
See #1884 which addresses your requested changes @raladev - would you mind approving this PR so we can merge these separately?

**Review by raladev — APPROVED** [review]

**Review by sangaman — CHANGES_REQUESTED** [review]
I thought we didn't want to mess with nodestateupdate packets in this PR? I was ready to merge this before the most recent commit. Let's try to keep the scope of PRs small and manageable, we can open separate PRs for separate issues.

**Comment by rsercano** [comment]
> I thought we didn't want to mess with nodestateupdate packets in this PR? I was ready to merge this before the most recent commit. Let's try to keep the scope of PRs small and manageable, we can open separate PRs for separate issues.

No no I didn't mean to mess with it, could you please check "files changed", I just was trying to sign the commits to make them mergeable.

**Review by raladev — APPROVED** [review]

**Comment by raladev** [comment]
> I thought we didn't want to mess with nodestateupdate packets in this PR? I was ready to merge this before the most recent commit. Let's try to keep the scope of PRs small and manageable, we can open separate PRs for separate issues.

822418a is just a revert of 7b7e7fc

**Comment by kilrau** [comment]
Merging is still blocked, because previous commits were not signed. Maybe you can squash commits into one first, then sign and force push to this branch. That should do the trick. @rsercano

**Comment by sangaman** [comment]
> No no I didn't mean to mess with it, could you please check "files changed", I just was trying to sign the previous commits to make them mergeable.

Got it, sorry that commit just confused me. So the current changes look good to me, if you can squash and sign these commits like kilrau said that would be great. Otherwise I can do it sometime tomorrow.

**Comment by rsercano** [comment]
I'm opening a new one https://github.com/ExchangeUnion/xud/pull/1890 sorry for the confusion, squashing looks like a pain since there're several commits between my commits, or Im literally too bad to do it!

**Pull request closed without merging**